src/components/implementation/no_interface/scheddev/sl_lock.c

```diff
@@ -0,0 +1,67 @@
+#include "sl.h"
+#include "sl_lock.h"
+
+void
+sl_lock_init(struct sl_lock *lock)
+{
+    assert(lock);
+    *lock = SL_LOCK_STATIC_INIT();
+}
+
+thdid_t
+sl_lock_holder(struct sl_lock *lock)
+{
+    assert(lock);
+    return lock->holder;
+}
+
+
+void
+sl_lock_lock(struct sl_lock *lock)
+{
+    assert(lock);
+
+    sl_cs_enter();
+    while (lock->holder != 0) {
+        sl_thd_yield_cs_exit(lock->holder);
+        sl_cs_enter();
+    }
+    lock->holder = sl_thdid();
+    sl_cs_exit();
+}
+
+int
+sl_lock_timed_lock(struct sl_lock *lock, microsec_t max_wait_time)
+{
+    int result;
+    cycles_t deadline = sl_now() + sl_usec2cyc(max_wait_time);
+
+    assert(lock);
+
+    sl_cs_enter();
+    while (lock->holder != 0 && sl_now() < deadline) {
+        sl_thd_yield_cs_exit(lock->holder);
+        sl_cs_enter();
+    }
+
+    if (lock->holder == 0) {
+        lock->holder = sl_thdid();
+        result = 1;
+    }else {
+        result = 0;
+    }
+    sl_cs_exit();
+    return result;
+}
+
+void
+sl_lock_unlock(struct sl_lock *lock)
+{
+    assert(lock);
+
+    sl_cs_enter();
+    assert(lock->holder);
```

Comment:
Assertions aren't there to help you debug.  They are there to fail early when there is a problem and give you a starting point for debugging.  They can be completely disabled when building, which points out their position as not really being a debugging aid.  I see no reason to incur overhead that is strictly redundant that all execution on the system will pay, to save yourself one recompile to add a few checks to see why it failed if it did.  That is not a good trade-off.